File: design_analysis_algorithms/question H algorithm analysis search times/algorithm_analysis_search_times.py
import time
import random

def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def measure_time(arr, target):
    start_time = time.time()
    binary_search(arr, target)
    return time.time() - start_time

# Create a sorted list for testing
arr = list(range(1, 1000000))

# Best Case: Target is the middle element
best_case_target = arr[len(arr) // 2]

# Worst Case: Target is not in the array
worst_case_target = -1

# Average Case: Random target from the array
average_case_target = random.choice(arr)

# Measure the execution time for each case
best_case_time = measure_time(arr, best_case_target)
worst_case_time = measure_time(arr, worst_case_target)
average_case_time = measure_time(arr, average_case_target)

print(f"Best case execution time: {best_case_time:.6f} seconds")
print(f"Worst case execution time: {worst_case_time:.6f} seconds")
print(f"Average case execution time: {average_case_time:.6f} seconds")

File: design_analysis_algorithms/question H algorithm analysis search times/algorithm_analysis_search_times.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <cstdlib>

using namespace std;
using namespace chrono;

int binary_search(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            return mid;
        }
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

void measure_time(const vector<int>& arr, int target) {
    auto start = high_resolution_clock::now();
    binary_search(arr, target);
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << "Execution time: " << duration.count() << " microseconds" << endl;
}

void analyzeSearchPerformance() {
    // Create a sorted array for testing
    vector<int> arr(1000000);
    for (int i = 0; i < 1000000; ++i) {
        arr[i] = i + 1;
    }

    // Best Case: Middle element
    int best_case_target = arr[arr.size() / 2];

    // Worst Case: Element not in array
    int worst_case_target = -1;

    // Average Case: Random element
    int average_case_target = arr[rand() % arr.size()];

    cout << "Best case scenario (target in middle): " << endl;
    measure_time(arr, best_case_target);

    cout << "Worst case scenario (target not in array): " << endl;
    measure_time(arr, worst_case_target);

    cout << "Average case scenario (random target): " << endl;
    measure_time(arr, average_case_target);
}

int main() {
    analyzeSearchPerformance();
    return 0;
}
